troubleshooting/tools/tracevol.py

```diff
@@ -362,28 +405,113 @@ def get_pool_name(arg, vol_id):
-    for pool in pools:
-        if int(pool_id) is int(pool['poolnum']):
-            return pool['poolname']
+        for pool in pools:
+            return pool['metadata_pool']
     return ""
 
+def check_subvol_in_cluster(arg, subVolName):
+    # check if user has specified subvolumeGroup
+    subvolumeGroup = get_subvolumeGroup(arg)
+    if subvolumeGroup == "":
+        # default subvolumeGroup
+        subvolumeGroup = "csi"
+    return check_subvol_path(arg, subVolName, subvolumeGroup)
+
+def check_subvol_path(arg, subVolName, subvolumeGroup):
+    cmd = ['ceph', 'fs', 'subvolume', 'getpath',
+           'myfs', subVolName, subvolumeGroup]
+    if not arg.userkey:
+        cmd += ["--id", arg.userid, "--key", arg.userkey]
+    if arg.toolboxdeployed is True:
+        tool_box_name = get_tool_box_pod_name(arg)
+        kube = [arg.command]
+        if arg.kubeconfig != "":
+            if arg.command == "oc":
+                kube += ["--config", arg.kubeconfig]
+            else:
+                kube += ["--kubeconfig", arg.kubeconfig]
+        kube += ['exec', '-it', tool_box_name, '-n',
+                 arg.rooknamespace, '--']
+        cmd = kube+cmd
+    out = subprocess.Popen(cmd, stdout=subprocess.PIPE,
+                           stderr=subprocess.STDOUT)
+    stdout, stderr = out.communicate()
+    if stderr is not None:
+        if arg.debug:
+            print("failed to get toolbox %s", stderr)
+        return False
+    if b"Error" in stdout:
+        if arg.debug:
+            print("subvolume not found in cluster", stdout)
+        return False
+    return True
+
+def get_subvolumeGroup(arg):
+    cmd = [arg.command]
+    if arg.kubeconfig != "":
+       if arg.command == "oc":
+          cmd += ["--config", arg.kubeconfig]
+       else:
+          cmd += ["--kubeconfig", arg.kubeconfig]
+    cmd += ['get', 'cm', arg.configmap, '-o', 'json']
+    out = subprocess.Popen(cmd, stdout=subprocess.PIPE,
+                           stderr=subprocess.STDOUT)
+    stdout, stderr = out.communicate()
+    if stderr is not None:
+        if arg.debug:
+            print("failed to get config map %s", stderr)
+            sys.exit()
+    try:
+        cm = json.loads(stdout)
+    except ValueError as err:
+        print(err, stdout)
+        sys.exit()
+    cmData = cm['data']['config.json']
+    if "subvolumeGroup" in cmData:
+        try:
+            cmDataJson = json.loads(cmData)
+        except ValueError as err:
+            print(err, stdout)
+            sys.exit()
+        for data in cmDataJson:
+            subvolumeGroup = data['cephFS']['subvolumeGroup']
+        return subvolumeGroup
+    else:
+        return ""
+
+def isRBDPV(myDict):
+    """
+    myDict contains key value store for annotations.
+    one of the annotations value is provisioner name.
+    provisioner name should tell whether pvc is for
+    cephfs or rbd.
+    """
+    for key in myDict:
+        hasRBD = "rbd" in myDict[key]
```

Comment:
Judging type by 'rbd' maybe not the best way because provisioner name can be configured.  How about the pv.Spec.CSI.VolumeAttributes["fsName"] ?

Reply:
Sure, I think I can also check if there is any volume attribute which has word "image" in the key then that PV will certainly belong to RBD.

Comments?

Reply:
Fixed